src/Toolchain.re

```diff
@@ -0,0 +1,759 @@
+             lsp: () => (name, [|"exec", "ocamllsp"|]),
+           }
+           |> R.return
+         });
+    };
+  };
+
+  let make = (~env, ~root, ~discoveredManifestPath, ~t) =>
+    switch (t) {
+    | Opam => Opam.make(~env, ~root, ~discoveredManifestPath)
+    | Esy => Esy.make(~env, ~root, ~discoveredManifestPath)
+    };
+
+  let toName =
+    fun
+    | Opam => "opam"
+    | Esy => "esy";
+
+  let ofName =
+    fun
+    | "opam" => Ok(Opam)
+    | "esy" => Ok(Esy)
+    | n => Error({j|Invalid name $n was provided|j});
+
+  let specOfName = (~env, ~name, ~root, ~discoveredManifestPath) =>
+    switch (name) {
+    | x when x == Opam.name => Opam.make(~env, ~root, ~discoveredManifestPath)
+    | x when x == Esy.name => Esy.make(~env, ~root, ~discoveredManifestPath)
+    | _ => "Invalid package manager name" |> R.fail |> Js.Promise.resolve
+    };
+
+  module PackageManagerSpecTuple = {
+    type t = (a, Fpath.t);
+    let compare = (x, y) => {
+      let (xa, xb) = x;
+      let (ya, yb) = y;
+      if (xa == ya && xb == yb) {
+        0;
+      } else {
+        1;
+      };
+    };
+  };
+
+  module PackageManagerSpecTupleSet = Set.Make(PackageManagerSpecTuple);
+
+  let alreadyUsed = folder => {
+    [Esy, Opam]
+    |> Array.fromList
+    |> Array.map(~f=pm => {
+         let lockFileFpath =
+           switch (pm) {
+           | Opam => Fpath.(join(folder, Opam.lockFile))
+           | Esy => Fpath.(join(folder, Esy.lockFile))
+           };
+         Fs.exists(Fpath.show(lockFileFpath))
+         |> Js.Promise.then_(Js.Promise.resolve << (exists => (pm, exists)));
+       })
+    |> Js.Promise.all
+    |> Js.Promise.then_(l =>
+         l
+         |> Array.filter(~f=((_pm, used)) => used)
+         |> Array.map(~f=t => {
+              let (pm, _used) = t;
+              pm;
+            })
+         |> Array.toList
+         |> R.return
+         |> Js.Promise.resolve
+       );
+  };
+  let available = (~env) => {
+    supportedPackageManagers
+    |> List.map(~f=(pm: t) => {
+         let name =
+           switch (pm) {
+           | Opam => Opam.name
+           | Esy => Esy.name
+           };
+         Cmd.make(~env, ~cmd=name)
+         |> Js.Promise.then_(
+              Js.Promise.resolve
+              << (
+                fun
+                | Ok(_) => {
+                    (pm, true);
+                  }
+                | Error(e) => {
+                    (pm, false);
+                  }
+              ),
+            );
+       })
+    |> Array.fromList
+    |> Js.Promise.all
+    |> Js.Promise.then_(
+         Js.Promise.resolve
+         << R.return
+         << Array.toList
+         << Array.map(~f=t => {
+              let (pm, _used) = t;
+              pm;
+            })
+         << Array.filter(~f=((_pm, available)) => available),
+       );
+  };
+
+  let setupToolChain = spec => spec.setup();
+  let lsp = spec => spec.lsp();
+  let env = spec => spec.env();
+
+  module Manifest: {
```

Comment:
Some comments explaining the terminology would be quite useful to me. I'm not entirely sure what is a manifest in this context for example